### PR TITLE
Raw specialization data

### PR DESCRIPTION
**Pull request by kvark**
Fixes a large portion of CTS "dEQP-VK.pipeline.spec_constant" group. The compute sub-part has 83 successes and 38 failures. The graphics part seems to suffer from a descriptor allocation issue, which is unrelated.
Related to https://github.com/gfx-rs/gfx/issues/2059#issuecomment-416439423

- [ ] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: metal
- [ ] `rustfmt` run on changed code

### CODE REVIEW

**Comment by kvark** [comment]
Hmm, actually now I wonder if a constant should just be `&[u8]` with an ID. This way we don't need to take care of a separate `data` array on HAL side, and it's not going to be any more complex on Vulkan backend side.

File: src/backend/gl/src/device.rs [hunk 1]
```diff
@@ -216,30 +216,25 @@ impl Device {
...
-                        pso::Constant::U32(v) => v as u64,
-                        pso::Constant::U64(v) => v,
-                        pso::Constant::I32(v) => *(&v as *const _ as *const u64),
-                        pso::Constant::I64(v) => *(&v as *const _ as *const u64),
-                        pso::Constant::F32(v) => *(&v as *const _ as *const u64),
-                        pso::Constant::F64(v) => *(&v as *const _ as *const u64),
-                    };
-                    ast.set_scalar_constant(spec_constant.id, value).map_err(gen_unexpected_error)?;
-                }
+                // Override specialization constant values
```
Comment (grovesNL) [inline src/backend/gl/src/device.rs]: This comment has been duplicated now

**Comment by grovesNL** [comment]
I guess we could still have an enum with `From` implemented or similar, if we wanted to provide some common type conversions.

**Comment by kvark** [comment]
Filed #2382 to follow up with a nicer API. Not seeing a clear way forward there, and the current state at least works.
bors r=grovesNL

**Comment by kvark** [comment]
CTS update: we pretty much nail down the specialization category of tests, but with a few caveats:
  - arrays are not supported for specialization constants (likely a SPIRV-Cross issue)
  - vertex shaders are generated with `void` return type (likely a SPIRV-Cross issue)
  - a workaround for #2383 is needed

**Pull request merged**